File: etsyCatalogs/etsyCatalogs/Model/CatalogServices.m
//
//  CatalogServices.m
//  etsyCatalogs
//
//  Created by Ihonahan Buitrago on 9/1/14.
//  Copyright (c) 2014 Ihonahan Buitrago. All rights reserved.
//

#import "CatalogServices.h"

@implementation CatalogServices


+ (void)getCatalogListingsByWeywords:(NSString *)keywords inPage:(int)nPage withBlock:(void (^)(CatalogResult *catalogResults, NSError *error))block
{
    // Set the Url and the request
    int pageOffset = kPageItemsize * nPage;
    NSString *url = [NSString stringWithFormat:@"%@%@", kServicesUrl, kListingService];
    url = [NSString stringWithFormat:url, kApiKey, keywords, pageOffset, kPageItemsize];
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
    
    // Call the Get method of the service
    [manager GET:url
      parameters:nil
         success:^(AFHTTPRequestOperation *operation, id responseObject) {
             if ([responseObject isKindOfClass:[NSDictionary class]]) {
                 NSDictionary *responseDict = (NSDictionary *)responseObject;
                 NSDictionary *pagination = [responseDict objectForKey:@"pagination"];
                 int count = [[responseDict objectForKey:@"count"] intValue];
                 if (count && pagination) {
                     // For us, a correct response has registries (possitive count) and pagination info
                     // So, we get the results array and create our catalog results object
                     CatalogResult *catalogResult = [[CatalogResult alloc] init];
                     [catalogResult insertPaginationInfo:pagination withCount:count];
                     
                     NSArray *results = [responseDict objectForKey:@"results"];
                     
                     for (NSDictionary *result in results) {
                         [catalogResult insertCatalogItem:result];
                     }
                     
                     // At the end, we execute the block passing it the catalogs result object
                     if (block) {
                         block(catalogResult, nil);
                     }
                 } else {   // If the results has no registries, we return an error reporting that
                     if (block) {
                         NSString *desc = NSLocalizedString(@"No results found", nil);
                         NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
                         NSError *noResultsError = [NSError errorWithDomain:kListingModelErrorDomain
                                                                          code:4020
                                                                      userInfo:userInfo];
                         block(nil, noResultsError);
                     }
                 }
             } else {   // We are expecting a specific dictionary, if the response is different then is an error
                 if (block) {
                     NSString *desc = NSLocalizedString(@"Bad response from server", nil);
                     NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
                     NSError *noJsonObjectError = [NSError errorWithDomain:kListingModelErrorDomain
                                                                      code:4001
                                                                  userInfo:userInfo];
                     block(nil, noJsonObjectError);
                 }
             }
         }
         failure:^(AFHTTPRequestOperation *operation, NSError *error) {
             if (block) {   // Bad request or something went wrong, we return an error
                 block(nil, error);
             }
         }];
}


+ (void)getListingImagesForListingID:(long long)listingId withBlock:(void (^)(NSArray *imagesResult, NSError *error))block
{
    // Set the Url and the request
    NSString *url = [NSString stringWithFormat:@"%@%@", kServicesUrl, kListingImagesService];
    url = [NSString stringWithFormat:url, listingId, kApiKey];
    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];

    // Call the Get method of the service
    [manager GET:url
      parameters:nil
         success:^(AFHTTPRequestOperation *operation, id responseObject) {
             if ([responseObject isKindOfClass:[NSDictionary class]]) {
                 NSDictionary *responseDict = (NSDictionary *)responseObject;
                 
                 int count = [[responseDict objectForKey:@"count"] intValue];
                 if (count) {
                     NSMutableArray *images = [NSMutableArray array];
                     NSDictionary *results = [responseDict objectForKey:@"results"];
                     for (NSDictionary *result in results) {
                         CatalogListingImage *image = [CatalogListingImage createCatalogListingImageFromDictionary:result];
                         [images addObject:image];
                     }
                     
                     // At the end, we execute the block passing it the images array we gathered from the
                     if (block) {
                         block(images, nil);
                     }
                 } else {
                     // If the results has no registries, we return an error reporting that
                     if (block) {
                         NSString *desc = NSLocalizedString(@"No results found", nil);
                         NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
                         NSError *noResultsError = [NSError errorWithDomain:kListingModelErrorDomain
                                                                       code:4020
                                                                   userInfo:userInfo];
                         block(nil, noResultsError);
                     }                 }
             } else {   // We are expecting a specific dictionary, if the response is different then its an error
                 if (block) {
                     NSString *desc = NSLocalizedString(@"Bad response from server", nil);
                     NSDictionary *userInfo = @{NSLocalizedDescriptionKey : desc};
                     NSError *noJsonObjectError = [NSError errorWithDomain:kListingModelErrorDomain
                                                                      code:4001
                                                                  userInfo:userInfo];
                     block(nil, noJsonObjectError);
                 }
             }
         }
         failure:^(AFHTTPRequestOperation *operation, NSError *error) {
             if (block) {   // Bad request or something went wrong, we return an error
                 block(nil, error);
             }
         }];
}

@end

File: etsyCatalogsSwift/etsyCatalogsSwift/Model/CatalogServices.swift
//
//  CatalogServices.swift
//  etsyCatalogsSwift
//
//  Created by Ihonahan Buitrago on 9/22/14.
//  Copyright (c) 2014 Ihonahan Buitrago. All rights reserved.
//

import Foundation


class CatalogServices : NSObject
{
    
    class func getCatalogListingsByKeywordsAndPage(keywords : String, pageNumber : Int, block : ((catalogsResults : CatalogResult?, error : NSError?) -> Void)! ) {
        if let responseBlock = block {
            let pageOffset : Int = ModelUtilities.kPageItemsize * pageNumber
            var url : String = ModelUtilities.kServicesUrl + ModelUtilities.kListingService
            url = String(format: url, ModelUtilities.kApiKey, keywords, pageOffset, ModelUtilities.kPageItemsize)
            var manager : AFHTTPRequestOperationManager = AFHTTPRequestOperationManager()
            
            manager.GET(url,
                parameters: nil,
                success: { (operation , responseObject) -> Void in
                    if let response : AnyObject = responseObject? {
                        if let responseDict = response as? Dictionary<String, AnyObject> {
                            var count : Int = responseDict["count"] as AnyObject? as Int
                            var pagination : Dictionary<String, AnyObject> = responseDict["pagination"] as AnyObject? as Dictionary<String, AnyObject>
                            
                            if count > 0 {
                                var catalogResult : CatalogResult = CatalogResult()
                                
                                catalogResult.insertPaginationInfo(pagination, count: count)
                                
                                var resultsArray : Array<AnyObject> = responseDict["results"] as AnyObject? as Array<AnyObject>
                                
                                for resultAny : AnyObject in resultsArray {
                                    var result : Dictionary<String, AnyObject> = resultAny as Dictionary<String, AnyObject>
                                    catalogResult.insertCatalogItem(result)
                                }
                                
                                block(catalogsResults: catalogResult, error: nil)
                            } else {
                                let givenError = CatalogServices.buildErrorWithCode(ModelUtilities.kErrorCodeNoResultsFound,
                                    description: NSLocalizedString("No results found", comment: ""))
                                block(catalogsResults: nil, error: givenError)
                            }
                        } else {
                            let givenError = CatalogServices.buildErrorWithCode(ModelUtilities.kErrorCodeBadResponse,
                                description: NSLocalizedString("Bad response from server", comment: ""))
                            block(catalogsResults: nil, error: givenError)
                        }
                    }
                },
                failure: { (operation, error) -> Void in
                    block(catalogsResults: nil, error: error)
            })
            
        }
    }
    
    
    class func getListingImagesForListingID(listingId : Int, block : ((imagesResult : Array<CatalogListingImage>?, error : NSError?) -> Void)! ) {
        if let responseBlock = block {
            var url : String = ModelUtilities.kServicesUrl + ModelUtilities.kListingImagesService
            url = String(format: url, listingId, ModelUtilities.kApiKey)
            var manager : AFHTTPRequestOperationManager = AFHTTPRequestOperationManager()
            
            manager.GET(url,
                parameters: nil,
                success: { (operation, responseObject) -> Void in
                    if let response : AnyObject = responseObject? {
                        if let responseDict = response as? Dictionary<String, AnyObject> {
                            var count : Int = responseDict["count"] as AnyObject? as Int
                            
                            if count > 0 {
                                var images : Array<CatalogListingImage> = []
                                var resultsArray : Array<AnyObject> = responseDict["results"] as AnyObject? as Array<AnyObject>
                                
                                for resultAny : AnyObject in resultsArray {
                                    var result : Dictionary<String, AnyObject> = resultAny as Dictionary<String, AnyObject>
                                    var item : CatalogListingImage = CatalogListingImage.createCatalogListingImageFromDictionary(result)
                                    images.append(item)
                                }
                                
                                block(imagesResult: images, error: nil)
                            } else {
                                let givenError = CatalogServices.buildErrorWithCode(ModelUtilities.kErrorCodeNoResultsFound,
                                    description: NSLocalizedString("No results found", comment: ""))
                                block(imagesResult: nil, error: givenError)
                            }
                        } else {
                            let givenError = CatalogServices.buildErrorWithCode(ModelUtilities.kErrorCodeBadResponse,
                                description: NSLocalizedString("Bad response from server", comment: ""))
                            block(imagesResult: nil, error: givenError)
                        }
                    }
                },
                failure: { (operation, error) -> Void in
                    block(imagesResult: nil, error: error)
            })
        }
    }
    
    
    
    
    
    // Mark: - Private methods
    private class func buildErrorWithCode(code : Int, description : String) -> NSError {
        let userInfo = [NSLocalizedDescriptionKey : description]
        var error : NSError = NSError(domain: ModelUtilities.kListingModelErrorDomain,
            code: code,
            userInfo: userInfo)
        return error
    }
    
}